src/esw/imu_driver.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python3
+import rospy
+from sensor_msgs.msg import Imu, MagneticField, Temperature
+from mrover.msg import CalibrationStatus
+
+import serial
+from serial import SerialException
+
+
+def main():
+    # publishers for all types of IMU data, queue size is 1 to make sure we don't publish old data
+    imu_pub = rospy.Publisher("imu", Imu, queue_size=1)
+    mag_pub = rospy.Publisher("magnetometer", MagneticField, queue_size=1)
+    temp_pub = rospy.Publisher("imu_temp", Temperature, queue_size=1)
+    calibration_pub = rospy.Publisher("imu_calibration", CalibrationStatus, queue_size=1)
+
+    # no rate used because rate is effectively set by Arduino serial publisher
+    rospy.init_node("imu_driver")
+
+    # read serial connection info and IMU frame from parameter server
+    port = rospy.get_param("imu_driver/port", "/dev/imu")
+    baud = rospy.get_param("imu_driver/baud", 115200)
+    imu_frame = rospy.get_param("imu_driver/frame_id", "imu_link")
+
+    # create serial connection with Arduino
+    ser = serial.Serial(port, baud)
+    attempts = 0
+
+    # initialize messages for IMU data
+    imu_msg = Imu()
+    mag_msg = MagneticField()
+    temp_msg = Temperature()
+    calibration_msg = CalibrationStatus()
+
+    while not rospy.is_shutdown():
+
+        # try to read a line from the serial connection,
+        # if this fails 100 times in a row then end the program
+        try:
+            line = ser.readline()
+            attempts = 0
+        except SerialException:
+            attempts += 1
+            if attempts > 100:
+                print("unable to read from serial port...ending program")
+                break
+
+        # parse data into a list of floats,
+        # if it fails to parse then skip this message
+        try:
+            data = [float(val.strip()) for val in line.split()]
+
+        except ValueError:
+            data = []
+            print("invalid msg format")
+            continue
+
+        # load parsed data into ROS messages,
+        # if it fails to get the data then skip this message
+        try:
+            # orientation
+            imu_msg.orientation.x = data[0]
```

Comment:
not your fault at all but this sucks. I wonder if we can shorten this with `getattr`

Reply:
check out my new solution

Reply:
New solution is goated btw